base/stats.go

```diff
@@ -0,0 +1,791 @@
+package base
+
+import (
+	"expvar"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	StatViewFormat = "%s.%s"
+
+	NamespaceKey                 = "sgw"
+	ResourceUtilizationSubsystem = "resource_utilization"
+
+	SubsystemCacheKey           = "cache"
+	SubsystemDatabaseKey        = "database"
+	SubsystemDeltaSyncKey       = "delta_sync"
+	SubsystemGSIViews           = "gsi_views"
+	SubsystemReplication        = "replication"
+	SubsystemReplicationPull    = "replication_pull"
+	SubsystemReplicationPush    = "replication_push"
+	SubsystemSecurity           = "security"
+	SubsystemSharedBucketImport = "shared_bucket_import"
+)
+
+type SgwStats struct {
+	GlobalStats     *GlobalStat         `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication,omitempty"`
+
+	mutex                     sync.Mutex
+	registeredReplicatorStats sync.Once
+}
+
+func NewSyncGatewayStats() *SgwStats {
+	sgwStats := SgwStats{
+		GlobalStats: &GlobalStat{},
+		DbStats:     map[string]*DbStats{},
+	}
+
+	sgwStats.GlobalStats.initResourceUtilizationStats()
+
+	return &sgwStats
+}
+
+func (s *SgwStats) String() string {
```

Comment:
I think it's worth mentioning in a comment that String() is here to satisfy the expvar.Var interface.